docs/smtchecker.rst

```diff
@@ -0,0 +1,733 @@
+parameters. This is enough to prove that the assertion above is always true.
+
+Abstracting a function call with an UF can be done for functions known to be
+deterministic, and can be easily done for pure functions.  It is however
+difficult to do this with general external functions, since they might depend
+on state variables.
+
+Reference Types and Aliasing
+============================
+
+Solidity implements aliasing for reference types with the same :ref:`data
+location<data-location>`.
+That means one variable may be modified through a reference to the same data
+area.
+The SMTChecker does not keep track of which references refer to the same data.
+This implies that whenever a local reference or state variable of reference
+type is assigned, all knowledge regarding variables of the same type and data
+location is erased.
+If the type is nested, the knowledge removal also includes all the prefix base
+types.
+
+::
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity >=0.8.0;
+    pragma abicoder v2;
+    pragma experimental SMTChecker;
+    // This will report a warning
+
+    contract Aliasing
+    {
+        uint[] array1;
+        uint[][] array2;
+        function f(
+            uint[] memory a,
+            uint[] memory b,
+            uint[][] memory c,
+            uint[] storage d
+        ) internal {
+            array1[0] = 42;
+            a[0] = 2;
+            c[0][0] = 2;
+            b[0] = 1;
+            // Erasing knowledge about memory references should not
+            // erase knowledge about state variables.
+            assert(array1[0] == 42);
+            // However, an assignment to a storage reference will erase
+            // storage knowledge accordingly.
+            d[0] = 2;
+            // Fails as false positive because of the assignment above.
+            assert(array1[0] == 42);
+            // Fails because `a == b` is possible.
+            assert(a[0] == 2);
+            // Fails because `c[i] == b` is possible.
+            assert(c[0][0] == 2);
+            assert(d[0] == 2);
+            assert(b[0] == 1);
+        }
+        function g(
+            uint[] memory a,
+            uint[] memory b,
+            uint[][] memory c,
+            uint x
+        ) public {
+            f(a, b, c, array2[x]);
+        }
+    }
+
+After the assignment to ``b[0]``, we need to clear knowledge about ``a`` since
+it has the same type (``uint[]``) and data location (memory).  We also need to
+clear knowledge about ``c``, since its base type is also a ``uint[]`` located
+in memory. This implies that some ``c[i]`` could refer to the same data as
+``b`` or ``a``.
+
+Notice that we do not clear knowledge about ``array`` and ``d`` because they
+are located in storage, even though they also have type ``uint[]``.  However,
+if ``d`` was assigned, we would need to clear knowledge about ``array`` and
+vice-versa.
+
+**********************
+Real World Assumptions
```

Comment:
Is this section ready? What is the actual (upper bound) assumption regarding arrays in the SMTChecker?

Reply:
This section was just copied from before, and wasn't changed at all by this PR. The assumption is that an array's length is always < type(uint256).max